mlkem/poly.c

```diff
@@ -512,20 +513,30 @@ void poly_tomont(poly *r) {
 /*************************************************
  * Name:        poly_reduce
  *
- * Description: Applies Barrett reduction to all coefficients of a polynomial
- *              for details of the Barrett reduction see comments in reduce.c
+ * Description: Converts polynomial to _unsigned canonical_ representatives.
+ *
+ *              The input coefficients can be arbitrary integers in int16_t.
+ *              The output coefficients are in [0,1,...,KYBER_Q-1].
  *
  * Arguments:   - poly *r: pointer to input/output polynomial
  **************************************************/
 #if !defined(MLKEM_USE_NATIVE_POLY_REDUCE)
 void poly_reduce(poly *r) {
   unsigned int i;
   for (i = 0; i < KYBER_N; i++) {
-    r->coeffs[i] = barrett_reduce(r->coeffs[i]);
+    // Barrett reduction, giving signed canonical representative
+    int16_t t = barrett_reduce(r->coeffs[i]);
+    // Conditional addition to get unsigned canonical representative
+    r->coeffs[i] = scalar_signed_to_unsigned_q_16(t);
```

Comment:
Isn't it easier to introduce an unsigned barrett reduction?

Reply:
you still need a conditional subtraction then, like in the avx2 code

Reply:
okay, then nvm.